source/common/http/header_map_impl.h

```diff
@@ -180,59 +188,89 @@ class HeaderMapImpl : NonCopyable {
    */
   class HeaderList : NonCopyable {
   public:
+    using HeaderNodeVector = absl::InlinedVector<HeaderNode, 1>;
+    using HeaderLazyMap = absl::flat_hash_map<absl::string_view, HeaderNodeVector>;
```

Comment:
it may be worth a few sentences here summarizing perf-tests that show it's better to use a flat_hash_map than a std::map here. I became enamored of std::map for this application, IIRC, because of the limited number of headers that are allowed in Envoy, and the performance curve I observed. There's a crossover point where flat_hash_map is better but it might be that the majority of cases are better off with std::map.

I may be remember wrong though.

Reply:
I've added the results of the benchmarks when using `std::map`. Although there isn't a clear winner, it seems that for thresholds 2-3 either `absl::flat_hash_map` is better or there isn't much difference.
If there are other things we should try, please let me know.

Reply:
oh...one other (I think I might have tried this earlier) use a std::multimap rather than a map<string_view, vector<HeaderHeaderNode>.

Reply:
No problem, I'll try `std::multimap`.

Reply:
The advantage of multimap and map is that you get fast implementation of prefix-matching, which I think maybe I had added a benchmark for in the old PR.

Reply:
Added usage of `std::multimap` and `absl::btree_multimap` (updated the comparison comment).
For most of the microbenchmarks there is almost no difference compared to `absl::flat_hash_map`, and in some cases the latter is much better.
For the countries benchmark, `std::mutimap` is about 2.5x slower, and `absl::btree_multimap` is about 1.5x slower compared to `absl::flat_hash_map`.
I'll look at the prefix-matching benchmark you added in the old PR.

Reply:
Thanks for working on this, this is awesome. It's surprisingly little code for the benefit so I feel good about the additional complexity we are adding for the benefit.

What is the final plan for how we will deploy this? Should we just make the threshhold runtime configurable? It could default at infinite for now if we want default off for initial testing or just pick a reasonable default based on some microbenchmarks of 5-10? If we are worried about perf we could just latch it once when the header map tables are created at startup? This would allow people to play with different thresholds on real systems or possible even tune for their use case?

My preference would be to get this change into a more final form without all the ifdefs and the runtime loading and then go from there? I don't have a strong opinion on flat_map vs. btree though I would prefer to just pick one and go with it. How do you suggest deciding what to go with?

One other thing to think about is whether we even want to support this functionality on trailers? Certainly some deployments may benefit from it but it seems very rare and I'm not sure if it's worth the extra cycles to initialize and check on every operation, though it's potentially mostly noise so no real strong opinion.

Cool stuff!

/wait

Reply:
Thanks for the comments!

> Thanks for working on this, this is awesome. It's surprisingly little code for the benefit so I feel good about the additional complexity we are adding for the benefit.
> 
> What is the final plan for how we will deploy this? Should we just make the threshhold runtime configurable? It could default at infinite for now if we want default off for initial testing or just pick a reasonable default based on some microbenchmarks of 5-10? If we are worried about perf we could just latch it once when the header map tables are created at startup? This would allow people to play with different thresholds on real systems or possible even tune for their use case?
> 
I think that making it runtime configurable, and loading it when a header map table is created could be costly. That said, I'll implement this to see how costly it is.
Regarding the latching at startup, what is the advantage of a runtime configuration over a bootstrap configuration value (if the value is only set once)?

> My preference would be to get this change into a more final form without all the ifdefs and the runtime loading and then go from there? I don't have a strong opinion on flat_map vs. btree though I would prefer to just pick one and go with it. How do you suggest deciding what to go with?
> 
Good question. I'd say it depends on the perf results. At the moment I lean towards flat_map, but if there's a use-case that is really useful and benefits from btree, then btree it is.
I'll add the prefix matching benchmark that @jmarantz suggested, and we'll see from there.
If there are other use-cases that we should test, please let me know.

> One other thing to think about is whether we even want to support this functionality on trailers? Certainly some deployments may benefit from it but it seems very rare and I'm not sure if it's worth the extra cycles to initialize and check on every operation, though it's potentially mostly noise so no real strong opinion.
> 
To keep it simple, all maps should be treated equal.
Another option is to have a threshold configuration for each of the map types, allowing the user to customize them to their use-cases.

> Cool stuff!
> 
> /wait

Reply:
> Regarding the latching at startup, what is the advantage of a runtime configuration over a bootstrap configuration value (if the value is only set once)?

There isn't much difference. Mostly just ability to tune without a config change. If we do put it bootstrap I would recommend using one of the `Runtime` types to still allow it to be runtime configurable. In general I think most users will not configure this so runtime seems fine to be but I don't feel strongly about it.

> Another option is to have a threshold configuration for each of the map types, allowing the user to customize them to their use-cases.

Agreed on keeping it simple. I like the idea of a different threshold for each map type. This shouldn't be very difficult and would be easy to add.

Reply:
I've added a runtime feature configuration that is being read and stored upon the first header-map creation.